File: kotlin-java-codingtest/src/main/kotlin/programmers/lv2/게임_맵_최단거리.java
package programmers.lv2;

import java.util.LinkedList;
import java.util.Queue;

public class 게임_맵_최단거리 {
    class Point {
        int x;
        int y;

        Point(int x, int y) {
            this.x = x;
            this.y = y;
        }

        public int getX() {
            return x;
        }

        public int getY() {
            return y;
        }

        public void setX(int x){
            this.x = x;
        }

        public void setY(int y) {
            this.y = y;
        }
    }

    public int solution(int[][] maps) {

        Queue<Point> q = new LinkedList<Point>();

        q.offer(new Point(0,0));

        int dx[] = {0,0,1,-1};
        int dy[] = {1,-1,0,0};

        int[][] visited = new int[maps.length][maps[0].length];

        visited[0][0] = 1;

        while(!q.isEmpty()) {
            Point target = q.poll();

            for(int i = 0; i<4; i++) {
                int nx = target.getX() + dx[i];
                int ny = target.getY() + dy[i];

                if(nx < 0 || ny < 0 || nx >= maps.length || ny >= maps[0].length || maps[nx][ny] == 0 || visited[nx][ny] > 0) continue;

                visited[nx][ny] = visited[target.getX()][target.getY()] + 1;

                if(nx == maps.length-1 && ny == maps[0].length-1) {
                    return visited[nx][ny];
                }

                q.offer(new Point(nx, ny));
            }
        }


        return -1;
    }
}

File: Python/programmers/lv2/게임 맵 최단거리.py
from collections import deque


def solution(maps):
    n = len(maps)
    m = len(maps[0])
    queue = deque()

    dx = [0, 0, -1, 1]
    dy = [-1, 1, 0, 0]

    visited = [[0] * m for _ in range(n)]
    visited[0][0] = 1
    queue.append((0, 0))

    while queue:
        (x, y) = queue.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < n and 0 <= ny < m and maps[nx][ny] == 1 and visited[nx][ny] == 0:
                visited[nx][ny] = visited[x][y] + 1
                queue.append((nx, ny))
                if nx == n - 1 and ny == m - 1:
                    return visited[nx][ny]

    return -1


print(solution([[1, 0, 1, 1, 1], [1, 0, 1, 0, 1], [1, 0, 1, 1, 1], [1, 1, 1, 0, 1], [0, 0, 0, 0, 1]]))
